src/shutdown/umount.c

```diff
@@ -716,10 +739,53 @@ static int mount_points_list_umount(MountPoint **head, bool *changed, bool last_
                         continue;
 
                 /* Trying to umount */
-                if (umount_with_timeout(m, last_try) < 0)
+                r = umount_with_timeout(m, last_try);
+                if (r < 0)
                         n_failed++;
                 else
                         *changed = true;
+
+                /* If a mount is busy, we move it to not keep parent mount points busy.
+                 * If a mount point is not a leaf, moving it would invalidate our mount table.
+                 * More moving will occur in next iteration with a fresh mount table.
+                 *
+                 * umount_with_timeout will return EPROTO even on EBUSY due to forking.
+                 */
+                if (r == -EPROTO && m->leaf) {
+                        _cleanup_free_ char *dirname = NULL;
+
+                        r = path_extract_directory(m->path, &dirname);
+                        if (r < 0) {
+                                n_failed++;
+                                log_full_errno(last_try ? LOG_ERR : LOG_INFO, errno, "Cannot find directory for %s: %m", m->path);
+                                continue;
+                        }
+
+                        if (!path_equal_or_files_same(dirname, "/run/shutdown/mounts", 0)) {
+                                _cleanup_free_ char *newpath = NULL, *hexhash = NULL;
+                                uint8_t hash[SHA256_DIGEST_SIZE];
+
+                                sha256_direct(m->path, strlen(m->path), hash);
+
+                                hexhash = hexmem(hash, SHA256_DIGEST_SIZE);
+                                if (!hexhash)
+                                        return log_oom();
+
+                                asprintf(&newpath, "/run/shutdown/mounts/%s", hexhash);
+                                if (!newpath)
+                                        return log_oom();
```

Comment:
That is both overcomplicated and not enough. It is possible to stack multiple mounts will be on the same path, which would result in the same hash and then we'd be stacking the mounts, but in reverse order. Just maintain a counter:
```C
char newpath[STRLEN("/run/shutdown/mounts/" + DECIMAL_STRING_MAX];
xsprintf(newpath, "/run/shutdown/mounts/%u", n_moved);
```

Reply:
counter would have to be static...

honestly I'd probably just use a random name. i.e. just use random_u64()

Reply:
It would be find if they got mounted in the same place. The important thing is that they do no get mounted in a subdirectory.

I can use a random number instead to make it simpler.

But using a hash was mainly to have reproducible error messages if there are failures to umount further. And it would allow, from a list of mount point, hash them and figure out which one is problematic. With random numbers or a counter, it will be very difficult to know what mount is problematic.